devito/arguments.py

```diff
@@ -1,68 +1,72 @@
 import abc
 
 import numpy as np
+from collections import OrderedDict
+from sympy import Symbol
 from cached_property import cached_property
-from collections import namedtuple
+from collections import defaultdict, namedtuple, OrderedDict
+from functools import reduce
+from itertools import chain
 
 from devito.exceptions import InvalidArgument
-from devito.logger import debug
-
-""" This module provides a set of classes that help in processing runtime arguments for
-    kernels generated by devito. There are two class hierarchies here:
-    - ArgumentProvider: These are for objects that might be used in the expression
-      provided to the operator as symbols but might resolve to runtime arguments after
-      code generation. Each ArgumentProvider provides one (or more) Argument
-      object(s).
-    - Argument: Classes inheriting from this are for objects that represent the
-      argument itself. Each ArgumentProvider might provide one or more such objects
-      which are used as placeholders for the argument as well as for verification and
-      derivation of default values.
+from devito.logger import debug, error
+from devito.tools import filter_ordered, flatten, GenericVisitor
+from devito.function import CompositeFunction, SymbolicFunction
+from devito.dimension import Dimension
+from devito.ir.support.stencil import retrieve_offsets
+
+
+""" This module contains a set of classes and functions to deal with runtime arguments
+to Operators. It represents the arguments and their relationships as a DAG (N, E) where
+every node (N) is represented by an object of class Parameter and every edge is an object
+of class Dependency. 
+The various class hierarchies are explained here:
+Parameter:
 """
 
-
-class Argument(object):
-
-    """ Abstract base class for any object that represents a run time argument for
-        generated kernels.
+class Parameter(object):
+    """ Abstract base class for any object that represents a node in the dependency
+        graph. It may or may not represent a runtime argument. 
     """
-
-    __metaclass__ = abc.ABCMeta
-
+    is_Argument = False
     is_ScalarArgument = False
     is_TensorArgument = False
     is_PtrArgument = False
-
-    def __init__(self, name, provider, default_value=None):
+    
+    __metaclass__ = abc.ABCMeta
+    def __init__(self, name, dependencies):
```

Comment:
I think it's important to provide a docstring here, especially for ``dependencies``

Reply:
Done